@azure/storage-common 12.4.0 browser bundle (dist/browser/crc64.js) contains require('fs'), breaking esbuild-based bundlers

### Description

`@azure/storage-common` 12.4.0 ships a browser bundle at `dist/browser/crc64.js`
that contains a `require('fs')` call (line 120). This is an emscripten-compiled
WASM wrapper that is a multi-environment build. The `require('fs')` is guarded by
`if (ENVIRONMENT_IS_NODE)` at runtime, but esbuild (and similar bundlers) resolve
`require()` calls **statically** regardless of surrounding conditionals.

This breaks any Angular 17+ project (or other esbuild-based builds) that depends
on `@azure/storage-blob`, which transitively depends on `@azure/storage-common`.

### Error

✘ [ERROR] Could not resolve "fs"

    node_modules/@azure/storage-common/dist/browser/crc64.js:120:23:
      120 │       var fs = require('fs');
          ╵                        ~~~~

  The package "fs" wasn't found on the file system but is built into node.
  Are you trying to bundle for node? You can use "platform: 'node'" to do that,
  which will remove this error.

### Affected versions

- `@azure/storage-common` 12.4.0 (broken)
- `@azure/storage-common` 12.3.0 (working)

### Root cause

`dist/browser/crc64.js` in 12.4.0 is a multi-environment emscripten build that
includes Node.js-specific code paths. The previous version (12.3.0) did not
have this issue.

### Workaround

Pin `@azure/storage-common` to `~12.3.0` via npm `overrides` in `package.json`:

```json
"overrides": {
  "@azure/storage-common": "~12.3.0"
}

Expected behavior

The browser bundle should not reference Node.js built-ins (fs, path) even
inside conditionals, since static bundlers cannot tree-shake require() calls.
```

--- a/sdk/storage/storage-common/CHANGELOG.md
+++ b/sdk/storage/storage-common/CHANGELOG.md
@@ -1,5 +1,11 @@
 # Release History
 
+## 12.4.1 (Unreleased)
+
+### Bugs Fixed
+
+- Fixed the browser and react-native builds of the CRC64 checksum calculator still containing Node.js `require('fs')`/`require('path')` calls, which broke esbuild-based bundlers. The post-build step now replaces the unreachable Node-only filesystem read block with a no-op in the browser and react-native copies of `crc64.js`. Issue [#38924](https://github.com/Azure/azure-sdk-for-js/issues/38924).
+
 ## 12.4.0 (2026-05-22)
 
 ### Features Added

--- a/sdk/storage/storage-common/copyJSFiles.cjs
+++ b/sdk/storage/storage-common/copyJSFiles.cjs
@@ -1,7 +1,7 @@
-const fs = require('fs');
+const fs = require("fs");
 
-const SRC_CRC64 = './src/crc64.js';
-const SRC_GLUE = './src/crc64_glue.js';
+const SRC_CRC64 = "./src/crc64.js";
+const SRC_GLUE = "./src/crc64_glue.js";
 
 // `dist/commonjs/package.json` is `{ "type": "commonjs" }`, so the same .js file
 // must use CommonJS syntax there. Strip the ESM-only polyfill block and rewrite
@@ -10,8 +10,21 @@ const SRC_GLUE = './src/crc64_glue.js';
 const ESM_COMPAT_BLOCK = /\/\/ ESM-COMPAT-START[\s\S]*?\/\/ ESM-COMPAT-END\r?\n?/;
 const ESM_EXPORT_BLOCK =
   /\/\/ ESM-EXPORT-START[^\n]*\r?\nexport default NativeCRC64;\r?\n\/\/ ESM-EXPORT-END/;
+// The Emscripten Node branch eagerly calls `require('fs')`/`require('path')` and wires
+// up filesystem-backed `read_`/`readBinary`/`readAsync`. Although this code is unreachable
+// in browsers (it is guarded by `ENVIRONMENT_IS_NODE`), bundlers such as esbuild statically
+// resolve the bare `require('fs')` calls and fail. Replace the block with a no-op for the
+// browser and react-native builds so no Node `require()` references remain.
+// See https://github.com/Azure/azure-sdk-for-js/issues/38924.
+const NODE_READ_BLOCK = /\/\/ NODE-READ-START[\s\S]*?\/\/ NODE-READ-END\r?\n?/;
+const NODE_READ_NOOP = `// Node-only filesystem reads omitted in browser/react-native builds.
+  read_ = readBinary = readAsync = () => {
+    throw new Error('NodeJS filesystem reads are not supported in this environment.');
+  };
 
-const esmSource = fs.readFileSync(SRC_CRC64, 'utf8');
+`;
+
+const esmSource = fs.readFileSync(SRC_CRC64, "utf8");
 
 if (!ESM_COMPAT_BLOCK.test(esmSource)) {
   throw new Error(
@@ -23,26 +36,30 @@ if (!ESM_EXPORT_BLOCK.test(esmSource)) {
     "copyJSFiles.cjs: expected ESM-EXPORT marker block in src/crc64.js was not found.",
   );
 }
+if (!NODE_READ_BLOCK.test(esmSource)) {
+  throw new Error(
+    "copyJSFiles.cjs: expected NODE-READ marker block in src/crc64.js was not found.",
+  );
+}
 
-fs.writeFileSync('./dist/esm/crc64.js', esmSource);
+fs.writeFileSync("./dist/esm/crc64.js", esmSource);
 
 const cjsSource = esmSource
-  .replace(
-    ESM_COMPAT_BLOCK,
-    "// ESM compatibility block omitted in CommonJS build.\n",
-  )
+  .replace(ESM_COMPAT_BLOCK, "// ESM compatibility block omitted in CommonJS build.\n")
   .replace(ESM_EXPORT_BLOCK, "module.exports = NativeCRC64;");
 
-fs.writeFileSync('./dist/commonjs/crc64.js', cjsSource);
+fs.writeFileSync("./dist/commonjs/crc64.js", cjsSource);
 
 const browserSource = esmSource
-      .replace(
-        ESM_COMPAT_BLOCK,
-        `// ESM compatibility block omitted in browsers build as it is NodeJS only.\n`);
+  .replace(
+    ESM_COMPAT_BLOCK,
+    `// ESM compatibility block omitted in browsers build as it is NodeJS only.\n`,
+  )
+  .replace(NODE_READ_BLOCK, NODE_READ_NOOP);
 
-fs.writeFileSync('./dist/browser/crc64.js', browserSource);
-fs.writeFileSync('./dist/react-native/crc64.js', browserSource);
+fs.writeFileSync("./dist/browser/crc64.js", browserSource);
+fs.writeFileSync("./dist/react-native/crc64.js", browserSource);
 
-for (const dir of ['browser', 'esm', 'commonjs', 'react-native']) {
+for (const dir of ["browser", "esm", "commonjs", "react-native"]) {
   fs.cpSync(SRC_GLUE, `./dist/${dir}/crc64_glue.js`);
 }

--- a/sdk/storage/storage-common/package.json
+++ b/sdk/storage/storage-common/package.json
@@ -3,7 +3,7 @@
   "sdk-type": "client",
   "sideEffects": false,
   "author": "Microsoft Corporation",
-  "version": "12.4.0",
+  "version": "12.4.1",
   "description": "Azure Storage Common Client Library for JavaScript",
   "license": "MIT",
   "repository": {

--- a/sdk/storage/storage-common/src/crc64.js
+++ b/sdk/storage/storage-common/src/crc64.js
@@ -133,6 +133,7 @@ function logExceptionOnExit(e) {
 
 if (ENVIRONMENT_IS_NODE) {
   if (typeof process == 'undefined' || !process.release || process.release.name !== 'node') throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');
+// NODE-READ-START (this block is replaced with a no-op in dist/browser and dist/react-native by copyJSFiles.cjs)
   // `require()` is no-op in an ESM module, use `createRequire()` to construct
   // the require()` function.  This is only necessary for multi-environment
   // builds, `-sENVIRONMENT=node` emits a static import declaration instead.
@@ -177,6 +178,7 @@ readAsync = (filename, onload, onerror) => {
 };
 
 // end include: node_shell_read.js
+// NODE-READ-END
   if (process['argv'].length > 1) {
     thisProgram = process['argv'][1].replace(/\\/g, '/');
   }

